src/Microsoft.CodeAnalysis.Analyzers/Core/ImplementationIsObsoleteAnalyzer.cs

```diff
@@ -0,0 +1,62 @@
+
+using System;
+using System.Collections.Immutable;
+using System.Linq;
+using Analyzer.Utilities;
+using Analyzer.Utilities.Extensions;
+using Microsoft.CodeAnalysis.Diagnostics;
+
+namespace Microsoft.CodeAnalysis.Analyzers
+{
+    using static CodeAnalysisDiagnosticsResources;
+
+    [DiagnosticAnalyzer(LanguageNames.CSharp, LanguageNames.VisualBasic)]
+    public class ImplementationIsObsoleteAnalyzer : DiagnosticAnalyzer
+    {
+        private const string ImplementationIsObsoleteAttributeName = "ImplementationIsObsoleteAttribute";
+        private const string ImplementationIsObsoleteAttributeFullName = "System.Runtime.CompilerServices.ImplementationIsObsoleteAttribute";
+
+        public static readonly DiagnosticDescriptor Rule = new(
+            DiagnosticIds.ImplementationIsObsoleteRuleId,
+            CreateLocalizableResourceString(nameof(ImplementationIsObsoleteTitle)),
+            CreateLocalizableResourceString(nameof(ImplementationIsObsoleteMessage)),
+            DiagnosticCategory.MicrosoftCodeAnalysisCompatibility,
+            DiagnosticSeverity.Error,
+            isEnabledByDefault: true,
+            description: CreateLocalizableResourceString(nameof(ImplementationIsObsoleteDescription)));
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics { get; } = ImmutableArray.Create(Rule);
+
+        public override void Initialize(AnalysisContext context)
+        {
+            context.EnableConcurrentExecution();
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+
+            context.RegisterSymbolAction(AnalyzeSymbol, SymbolKind.NamedType);
+        }
+
+        private static void AnalyzeSymbol(SymbolAnalysisContext context)
+        {
+            var namedTypeSymbol = (INamedTypeSymbol)context.Symbol;
+
+            // If any interface implemented by this type has the attribute and if the interface and this type are not
+            // in "internals visible" context, then issue an error.
+            foreach (INamedTypeSymbol iface in namedTypeSymbol.AllInterfaces)
+            {
+                System.Collections.Generic.IEnumerable<AttributeData> attributes = iface.GetAttributes();
+
+                // We are doing a string comparison of the name here because we don't care where the attribute comes from.
+                // CodeAnalysis.dll itself has this attribute and if the user assembly also had it, symbol equality will fail
+                // but we should still issue the error.
+                if (attributes.FirstOrDefault(a => a is { AttributeClass.Name: ImplementationIsObsoleteAttributeName, ConstructorArguments: [{ Value: string }] }
+                                                   && a.AttributeClass.ToDisplayString().Equals(ImplementationIsObsoleteAttributeFullName, StringComparison.Ordinal)) is { } attr &&
+                    !iface.ContainingAssembly.GivesAccessTo(namedTypeSymbol.ContainingAssembly))
```

Comment:
>!iface.ContainingAssembly.GivesAccessTo(namedTypeSymbol.ContainingAssembly) [](http://example.com/codeflow?start=20&length=75)

What would be the downside of removing this access check? Would the analyzer be too noisy/annoying?

Reply:
It would warn in our unit tests, for example.